### ISSUE TITLE
Comparison of `std::optional<not_null<T>>` and `not_null<T>`  is ambiguous

### ISSUE THREAD

**Issue by MatthewNielsen27** [gsl-lite/gsl-lite#346]
The comparison of `std::optional<not_null<T>>` and `not_null<T>` is ambiguous. This is possibly because they both have overloads of the equality operator on their "nullable" types, which are of equal precedence. 

Here's a minimal example to repro:  https://godbolt.org/z/xnEfP1acb

It should be noted that `Microsoft/GSL` doesn't have this issue: https://godbolt.org/z/Ed7qG5EE8

**Comment by mbeutel** [comment]
Thanks for the report.

I have yet come up with a convincing approach to solve this problem; suggestions are welcome. We might be tempted to follow MS-GSL and only define comparison operators for two `not_null<>` arguments ([link](https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/pointers#L162-L168)) rather than the set of mixed-argument overloads we have now ([link](https://github.com/gsl-lite/gsl-lite/blob/db7a9c9108f857506608f8012f00e51ff8edbf50/include/gsl/gsl-lite.hpp#L3336-L3356)); but the simpler approach is fraught with other problems: for example, comparing `std::shared_ptr<T>` and `not_null<std::shared_ptr<T>>` then fails, whereas it works fine with *gsl-lite*. (Repro for [MS-GSL](https://godbolt.org/z/zbPG3eTa4), [*gsl-lite*](https://godbolt.org/z/65h36Wccq).)

As an aside, isn't `optional<not_null<T>>` just a contrived semantic equivalent of `T`?

**Comment by MatthewNielsen27** [comment]
> As an aside, isn't optional<not_null<T>> just a contrived semantic equivalent of T?

Yeah, kinda. But when designing generic types / interfaces with value semantics, this issue adds some mental overhead. The pattern is used frequently in Rust, with `Option<Box<T>>` which is equivalently `optional<not_null<T*>>`.
